api/commands/run.ts

```diff
@@ -96,107 +100,84 @@ export class Runner {
-        // All required deps are completed, start this action.
-        this.executeAction(action);
-      } else if (this.cancelled || finishedDeps.length == action.dependencies.length) {
-        await this.triggerChange();
-        // All deps are finished but they weren't all successful, or the run was cancelled.
-        // skip this action.
-        this.result.actions.push({
-          name: action.name,
-          status: dataform.ActionExecutionStatus.SKIPPED,
-          deprecatedSkipped: true
-        });
-      } else {
-        this.pendingActions.push(action);
-      }
+  private async executeGraph() {
+    const executeAllActionsReadyForExecution = async () => {
+      const allSuccessfulActions = this.result.actions
+        .filter(isSuccessfulAction)
+        .map(fn => fn.name);
+      const isReadyForExecution = (action: dataform.IExecutionAction) => {
+        for (const dependency of action.dependencies) {
+          if (!allSuccessfulActions.includes(dependency)) {
+            return false;
+          }
+        }
+        return true;
+      };
+      const readyForExecutionActions = this.pendingActions.filter(isReadyForExecution);
+      this.pendingActions = this.pendingActions.filter(action => !isReadyForExecution(action));
+      return Promise.all(
+        readyForExecutionActions.map(async action => {
+          this.result.actions.push(await this.executeAction(action));
+          await this.triggerChange();
+          await executeAllActionsReadyForExecution();
+        })
+      );
+    };
+    // Recursively execute all actions as they become executable.
+    await executeAllActionsReadyForExecution();
+
+    let ok = true;
+    this.result.actions.forEach(action => {
+      ok = ok && isSuccessfulAction(action);
     });
+    this.result.ok = ok;
+  }
 
-    if (this.pendingActions.length > 0 || this.result.actions.length != this.graph.actions.length) {
-      setTimeout(() => this.loop(resolve, reject), 100);
-    } else {
-      // Work out if this run was an overall success.
-      let ok = true;
-      this.result.actions.forEach(action => {
-        ok =
-          ok &&
-          (action.status === dataform.ActionExecutionStatus.SUCCESSFUL ||
-            action.status == dataform.ActionExecutionStatus.DISABLED);
-      });
-      this.result.ok = ok;
-      resolve();
+  private async executeAction(action: dataform.IExecutionAction) {
+    const startTime = process.hrtime();
+
+    const executedTasks: dataform.IExecutedTask[] = [];
+    let allSuccessful = true;
+    for (const task of action.tasks) {
+      if (allSuccessful) {
+        const executedTask = await this.executeTask(task);
+        executedTasks.push(executedTask);
+        allSuccessful = allSuccessful && executedTask.ok;
+      }
     }
+
+    const endTime = process.hrtime(startTime);
+    const executionTime = endTime[0] * 1000 + Math.round(endTime[1] / 1000000);
```

Comment:
This is super weird (I know it was already there). To get timestamp millis, just do new Date().valueOf().
(doubles are a safe way to represent milli timestamps until the year 2255 AIUI, although the above code doesn't even really circumvent that).